File: frontend/src/api/locationService.ts
import { api } from './client'

export interface Location {
  id: number
  name: string
  latitude: number
  longitude: number
}

export interface PageResponse<T> {
  content: T[]
  totalElements: number
  totalPages: number
  size: number
  number: number
}

export const locationService = {
  getAllLocations: async (page = 0, size = 10) => {
    const response = await api.get<PageResponse<Location>>(`locations?page=${page}&size=${size}`)
    return response.data
  },

  getLocationById: async (id: number) => {
    const response = await api.get<Location>(`locations/${id}`)
    return response.data
  },

  createLocation: async (location: Omit<Location, 'id'>) => {
    const response = await api.post<Location>('locations', location)
    return response.data
  },

  updateLocation: async (id: number, location: Omit<Location, 'id'>) => {
    const response = await api.put<Location>(`locations/${id}`, location)
    return response.data
  },

  deleteLocation: async (id: number) => {
    await api.delete(`locations/${id}`)
  }
} 

File: backend/src/main/java/com/furkanbegen/routes/service/LocationService.java
package com.furkanbegen.routes.service;

import com.furkanbegen.routes.dto.LocationDTO;
import com.furkanbegen.routes.exception.*;
import com.furkanbegen.routes.mapper.LocationMapper;
import com.furkanbegen.routes.repository.LocationRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class LocationService {

  private final LocationRepository locationRepository;
  private final LocationMapper locationMapper;

  public Page<LocationDTO> getAllLocations(Pageable pageable) {
    return locationRepository.findAll(pageable).map(locationMapper::toDTO);
  }

  public LocationDTO getLocationById(Long id) {
    var location =
        locationRepository
            .findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Location not found"));

    return locationMapper.toDTO(location);
  }

  public LocationDTO createLocation(LocationDTO dto) {

    var location = locationMapper.toEntity(dto);

    return locationMapper.toDTO(locationRepository.save(location));
  }

  public LocationDTO updateLocation(Long id, LocationDTO dto) {

    var location =
        locationRepository
            .findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Location not found"));

    var locationForUpdate = locationMapper.toEntity(dto, location);

    return locationMapper.toDTO(locationRepository.save(locationForUpdate));
  }

  public void deleteLocation(Long id) {
    var location =
        locationRepository
            .findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Location not found"));

    locationRepository.delete(location);
  }

  public Page<LocationDTO> searchLocations(String query, Pageable pageable) {
    return locationRepository.searchByNameContainingIgnoreCase(query, pageable)
        .map(locationMapper::toDTO);
  }
}
